Generate Operations Repository Payload

aquarat opened the pull request:
This PR is mostly to add a script which generates an operations repository payload. An example of what it generates is in this PR: https://github.com/api3dao/operations/pull/3 (which is from the recent Amberdata deployment)

The biggest issue is the monitoring wallet address generation (see line 195 in `export-for-ops.ts`).
I think we need to monitor these wallets operationally:
1. The provider's Airnode-controlled wallet, used to fulfill requests via the AirnodeRrp contract instance.
2. The provider's keeper-controlled wallet, used to request updates via the Beacon Server contract instance.
3. The above two wallets belonging to API3 for backup services.

Additionally, to aid documentation in future the script also produces two documentation-specific outputs. These were guided by what @wkande asked for, although can be trimmed.

The original operations repository schema does not call for airkeeper.json to be included, but does include Airnode's config.json, so I have taken the liberty of including Airkeeper's config file too. I haven't included the secrets file.

aquarat committed:
Initial documentation and operations payload exporter

aquarat committed:
Re-add operations repository exporter and fix Husky.

aquarat committed:
Add a chains.json output and do some polish.

aquarat committed:
Modify config path

aquarat committed:
Fix beaconId

acenolaza commented on scripts/export-for-ops.ts:
```diff
@@ -0,0 +1,249 @@
+                templateObj.airnode,
+                requestSponsor
+              ),
+              api3Unknown: await deriveSponsorWalletAddress(api3Xpub, api3AirnodeAddress, requestSponsor),
```
Why is this key `api3Unknown` instead of `api3SponsorWallet` or something similar?

aquarat replied:
This is what I was hoping you could clarify :)
There are three wallets defined in the original operations repository schema:
`sponsor`
`apiProviderAirkeeperSponsor`
`api3AirkeeperSponsor`

This terminology is very confusing to me. I have renamed them, for now, to:
`api3Unknwon` - I'm not sure what this wallet does or where it comes from. Is it API3's master wallet used to transfer funds to target wallets?
`apiProviderAirnodeFulfiller` - this is the Airnode-controlled wallet that executes fulfillments
`apiProviderUpdateRequester` - this is the Airkeeper-controlled wallet that calls for updates via the Beacon Server contract.

I have based this on the wallets I think are important/the wallets we want to keep funded; these wallets seem to be the only wallets that experience gas-token usage.

What do you think? Please feel free to think broadly/not be constrained by what's already present, both here and in the operations repository.
And once this is done I'll update the operations repository README.

aquarat replied:
I've renamed the addresses - no one has been able to give me definitive guidance on the naming so far, so clearly the language is a problem generally.

We can rename the addresses at a later stage if necessary.

Ashar2shahid replied:
I don't think we should be renaming the addresses, 
i'll clear it up for you:

sponsor : Sponsor is an EAO account that will have an associated `sponsorWallet` that is derived from the airnode mnemonic. The sponsor specifies the requester contract that the `sponsorWallet` can fulfill on behalf of. In the case of the airkeeper the sponsor can be the airnode address itself since the airnode is sponsoring the RrpBeaconServer.

apiProviderAirkeeperSponsor : Similar to how sponsor works, there are also keeper sponsors. The `KeeperSponsorWallet` is derived from the keeper sponsor. This is the address that sends the `request beacon update` to the RrpBeaconServer after checking the deviation. 

Since there are plans to have two airkeepers running, one by the api provider and the other by api3(at a higher threshold)for every beacon, we will have two keeper sponsors aka the `apiProviderAirkeeperSponsor` and `api3AirkeeperSponsor` which will have their own KeeperSponsorWallets respectively.

aquarat replied:
> I don't think we should be renaming the addresses, i'll clear it up for you:
> 
> sponsor : Sponsor is an EAO account that will have an associated `sponsorWallet` that is derived from the airnode mnemonic. The sponsor specifies the requester contract that the `sponsorWallet` can fulfill on behalf of. In the case of the airkeeper the sponsor can be the airnode address itself since the airnode is sponsoring the RrpBeaconServer.
> 
> apiProviderAirkeeperSponsor : Similar to how sponsor works, there are also keeper sponsors. The `KeeperSponsorWallet` is derived from the keeper sponsor. This is the address that sends the `request beacon update` to the RrpBeaconServer after checking the deviation.
> 
> Since there are plans to have two airkeepers running, one by the api provider and the other by api3(at a higher threshold)for every beacon, we will have two keeper sponsors aka the `apiProviderAirkeeperSponsor` and `api3AirkeeperSponsor` which will have their own KeeperSponsorWallets respectively.

Thanks @Ashar2shahid, that helps a lot, I will make the necessary changes tomorrow in a new PR:)

aquarat replied:
@Ashar2shahid This is what I'm after, please define these for me:

- sponsor: Sponsor address that will be used while requesting an update (sponsor wallet to be derived from the API provider xpub). Both the API provider-operated and the API3-operated Airkeepers use the same sponsor.
- * Please confirm: This is a key that isn't actually transacted on from which other wallets are derived?


- apiProviderAirkeeperSponsor: Airkeeper sponsor address that will be used by the API provider-operated Airkeeper to keep the beacon (sponsor wallet to be derived from xpub in apiMetadata.json)
- * Please confirm: This wallet is used by the API Provider's Airkeeper to request an update?


- api3AirkeeperSponsor: Airkeeper sponsor address that will be used by the API3-operated Airkeeper to keep the beacon (sponsor wallet to be derived from xpub in api3Metadata.json)
- * Please confirm: this is a wallet used by API3's Airkeeper to request an update as a failsafe? and derived like so (line 185 in export-for-ops):
- * `deriveSponsorWalletAddress(api3Xpub, api3AirnodeAddress, requestSponsor)`


- ?????????: The API Provider's Airnode-controlled wallet used to fulfill requests?

Operationally we need to know what wallets to top up, so this would be API3's Airkeeper update requester wallet, the API Provider's Airkeeper-controlled wallet and the API Provider's Airnode-controlled fulfiller wallet.

Separately, have you done any work around naming templates?

Thanks :)

acenolaza replied:
Maybe I can help but first I'd like to make sure we all understand this:

Sponsoring means that there's an address (sponsor EOA probably) that will be used to derive a path for Airnode/Airkeeper to generate wallets based on that path in order to make transactions. These are called "sponsorWallets" and are the ones that need to be funded.

> sponsor: Sponsor address that will be used while requesting an update (sponsor wallet to be derived from the API provider xpub). Both the API provider-operated and the API3-operated Airkeepers use the same sponsor.
> 
> Please confirm: This is a key that isn't actually transacted on from which other wallets are derived?

This address is the one Airnode will use to **derive the path** of a wallet that is under its control in order to fulfill requests. This wallet is called the [sponsorWallet](https://github.com/api3dao/operations/blob/main/src/dump-wallet-addresses.ts#L34-L44) and needs to be funded.

> apiProviderAirkeeperSponsor: Airkeeper sponsor address that will be used by the API provider-operated Airkeeper to keep the beacon (sponsor wallet to be derived from xpub in apiMetadata.json)
> 
> Please confirm: This wallet is used by the API Provider's Airkeeper to request an update?

This address will be used by the API provider operated Airkeeper to **derive the path** when creating  a `sponsorWallet` which is actually the one used to send beacon udpate transactions. This [sponsorWallet](https://github.com/api3dao/operations/blob/main/src/dump-wallet-addresses.ts#L45-L55) also needs to be funded.

> api3AirkeeperSponsor: Airkeeper sponsor address that will be used by the API3-operated Airkeeper to keep the beacon (sponsor wallet to be derived from xpub in api3Metadata.json)
> 
> Please confirm: this is a wallet used by API3's Airkeeper to request an update as a failsafe? and derived like so (line 185 in export-for-ops):
> deriveSponsorWalletAddress(api3Xpub, api3AirnodeAddress, requestSponsor)

This address will be used by the API3 operated Airkeeper to **derive the path** when creating  a `sponsorWallet` which is actually the one used to send beacon udpate transactions (as a failsafe like you mentioned with a higher threshold). This [sponsorWallet](https://github.com/api3dao/operations/blob/main/src/dump-wallet-addresses.ts#L56-L67) also needs to be funded.

> ?????????: The API Provider's Airnode-controlled wallet used to fulfill requests?

This will be the [sponsorWallet](https://github.com/api3dao/operations/blob/main/src/dump-wallet-addresses.ts#L34-L44) derived using the sponsor address 

> Operationally we need to know what wallets to top up, so this would be API3's Airkeeper update requester wallet, the API Provider's Airkeeper-controlled wallet and the API Provider's Airnode-controlled fulfiller wallet.

This is correct. More accurately all three [sponsorWallets](https://github.com/api3dao/operations/blob/main/src/dump-wallet-addresses.ts) listed here.

acenolaza replied:
BTW I just realized that [we have changed](https://github.com/api3dao/operations/pull/2) the `protocolId` section of the derivation path for v0.4 but we have decided to stick to v0.3.1 so [this script](https://github.com/api3dao/operations/blob/main/src/dump-wallet-addresses.ts#L37) might generate a wrong sponsorWallet address path for the hackathon.

Siegrift replied:
@acenolaza You are probably right. I've asked this https://api3dao.atlassian.net/browse/AN-453?focusedCommentId=12976 but I should have messaged the beacon channel to be careful about it...

aquarat replied:
Thanks @acenolaza that's great, exactly what I was after. I will make the changes.

acenolaza replied:
> @acenolaza You are probably right. I've asked this https://api3dao.atlassian.net/browse/AN-453?focusedCommentId=12976 but I should have messaged the beacon channel to be careful about it...

@Siegrift sorry I missed that message but if we plan to use the script with v0.3.1 then we should revert the change or create a branch with a revert commit only for hackathon like you suggested [here](https://api3workspace.slack.com/archives/C02UW1M157A/p1643374411080499).

acenolaza reviewed (approved):
LGTM 🥇

aquarat committed:
Various improvements
- Exported data is now tar.gzipped
- apiName is asked for during setup-integration
- Data provider contact email is now asked for during setup-integration and
- provided details are included in export-for-ops payload
- Improvements to operations schema
- TODO README updates
